core/src/main/java/com/xmartlabs/suspendedlocation/core/SuspendedLocationImpl.kt

```diff
@@ -0,0 +1,220 @@
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.io.IOException
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+internal class SuspendedLocationImpl(
+    internal var context: Context,
+    internal var locationServices: FusedLocationProviderClient) {
+
+  /**
+   * A function that returns a [Flow] that emits your current [Location] given a [LocationRequest]
+   *
+   * @param locationRequest The [LocationRequest] to be called.
+   * @param runnerContext The context in which the function will be executed.
+   *
+   * @throws IOException if the location is not available
+   */
+  @ExperimentalCoroutinesApi
+  @RequiresPermission(
+      anyOf = ["android.permission.ACCESS_COARSE_LOCATION", "android.permission.ACCESS_FINE_LOCATION"]
+  )
+  internal fun requestCurrentLocation(
+      locationRequest: LocationRequest,
+      runnerContext: CoroutineContext = Dispatchers.Default
+  ): Flow<Location> = callbackFlow<Location> {
+    locationServices
+        .requestLocationUpdates(
+            locationRequest,
+            object : LocationCallback() {
+              @Synchronized
+              override fun onLocationResult(locationResult: LocationResult) {
+                launch(runnerContext) {
+                  this@callbackFlow.send(locationResult.lastLocation)
+                }
+              }
+
+              override fun onLocationAvailability(locationAvailability: LocationAvailability) {
+                super.onLocationAvailability(locationAvailability)
+                if (!locationAvailability.isLocationAvailable) {
+                  launch(runnerContext) {
+                    this@callbackFlow.cancel(
+                        "Location not available",
+                        IOException("Location not available")
+                    )
+                  }
+                }
+              }
+            },
+            locationServices.looper
+        )
+    awaitClose { cancel() }
+  }.flowOn(runnerContext)
+
+  /**
+   * A function that returns your current [Location]
+   *
+   * @param runnerContext The context in which the function will be executed.
+   *
+   * @throws IOException if the location is not available
+   */
+  @RequiresPermission(
+      anyOf = ["android.permission.ACCESS_COARSE_LOCATION", "android.permission.ACCESS_FINE_LOCATION"]
+  )
+  internal suspend fun requestCurrentLocation(
+      runnerContext: CoroutineContext = Dispatchers.Default
+  ): Location = withContext(runnerContext) {
+    suspendCoroutine { continuation: Continuation<Location> ->
+      locationServices
+          .requestLocationUpdates(
+              LocationRequest().setInterval(1),
+              object : LocationCallback() {
+                override fun onLocationResult(locationResult: LocationResult) {
```

Comment:
maybe in another pr, but I'd refactor it. I'd use flow here

Reply:
we allowed both one with coroutines and other with flow